### PR TITLE
Refactor `torch.Generator` bindings to use pybind11

### PR DESCRIPTION
**Pull request by lqf96**
Refactor `torch.Generator` bindings to use pybind11, as Python bindings with pybind11 do not need to deal with low-level Python APIs and reference counting, and thus the code is much cleaner. This is a preparation for implementing pickling support for `torch.Generator` (#43672). Depends on #49589.

## What is being refactored?

* Interfaces
  - `{c10::GeneratorImpl, at::Generator}::{pyobj, set_pyobj}` are removed, as they are no longer needed.
* PyTorch Bindings
  - Class `THPGenerator` and all `THPGenerator_*` functions are removed. `torch.Generator` now directly binds to `at::Generator` through pybind11 in `torch::initGeneratorBindings`.
  - Instance checks now use `py::isinstance<at::Generator>(...)` instead of `THPGenerator_Check`.
  - `py::cast` is used to convert between `at::Generator` and `torch.Generator`.
* Utilities
  - `at::Generator` pybind11 type caster is removed, as it is now directly bound to Python.
  - `THPModule_addDocStr` now handles `PyInstanceMethod` generated by pybind11 for class methods. Note that the implementation is extremely hacky and uses implementation details of pybind11.
  - A python `add_docstr` implementation is created and handles docstring addition for pybind11 types and properties. For any other objects, it simply forwards to `torch._C._add_docstr` (`THPModule_addDocStr`).
* Fixed docstring tests and an exception test when an invalid seed is provided to the Generator.

## Questions

~~As mentioned in issue #49797, this pull request introduces pybind11 to core PyTorch APIs and removes `THP*` functions. Thus, it needs to be clarified whether these APIs are considered implementation details or public APIs. Otherwise merging this pull request will break backward compatibility for certain C extensions.~~

It has been clarified that `THP*` functions are internal APIs and migrating them to pybind11 is OK as long as it does not cause performance issues.

### CODE REVIEW

**Comment by lqf96** [comment]
This is now ready for review. cc @ngimel and @pbelevich since both of you are assigned to the previous pull request.

**Comment by pbelevich** [comment]
I will review it after https://github.com/pytorch/pytorch/pull/49589 merge to master since it looks like it needs to be rebased
@lqf96 please consider using https://github.com/ezyang/ghstack if you want stacked PRs

**Review by pbelevich — COMMENTED** [review]
@lqf96 Can we skip `py::is_final` flag? Maybe fbcode's intenal version is old and doesn't have it

File: torch/csrc/Generator.cpp [hunk 1]
```diff
@@ -1,272 +1,125 @@
...
-  TORCH_CHECK(self.defined(), "cannot call set_pyobj() on undefined generator");
-  self.set_pyobj(pyobj);
+} // namespace
+
+void initGeneratorBindings(PyObject* module) {
+  py::options options;
+  options.disable_user_defined_docstrings();
+  options.disable_function_signatures();
+
+  py::class_<Generator>(module, "Generator", py::is_final())
```
[the remark this answers is not in the document]
Comment (lqf96) [reply torch/csrc/Generator.cpp:49]: Yeah `py::is_final` is [a fairly recent addition](https://github.com/pybind/pybind11/releases/tag/v2.6.0). I think we can at least skip it here since in the current version of PyTorch it's possible to subclass `torch.Generator` in Python, but I'll need to come up with a solution for other binding classes which disallow it (e.g. `torch.device`, `torch.dtype`).

[the remark this answers is not in the document]
Comment (lqf96) [reply torch/csrc/Generator.cpp]: Done.

**Comment by lqf96** [comment]
There are some weird lint errors coming from `torch/csrc/utils.cpp`:

```
[bugprone-suspicious-include] warning: suspicious #include of file with '.cpp' extension
[cppcoreguidelines-avoid-c-arrays,modernize-avoid-c-arrays] warning: do not declare C-style arrays, use std::array<> instead
...
```

and I have no idea why they're showing up :|

File: torch/csrc/Module.cpp [hunk 2]
```diff
@@ -267,6 +265,34 @@ PyObject *THPModule_addDocStr(PyObject *_unused, PyObject *args)
...
+      if (strlen(func->m_ml->ml_doc) != 0) {
+        return PyErr_Format(PyExc_RuntimeError,
+            "pybind11 method '%s' already has a docstring "
+            "(to fix this error, disable pybind11's automatic docstring generation)",
+            func->m_ml->ml_name);
+      }
+      // This is extremely hacky; it uses the implementation detail of pybind11 that
+      // docstrings are allocated with `malloc()`
+      //NOLINTNEXTLINE(cppcoreguidelines-no-malloc, cppcoreguidelines-pro-type-const-cast)
+      free(const_cast<char *>(func->m_ml->ml_doc));
```
Comment (pbelevich) [inline torch/csrc/Module.cpp:289]: @ezyang WDYT? I need second opinion

**Comment by pbelevich** [comment]
> There are some weird lint errors coming from `torch/csrc/utils.cpp`:

It's not introduced by you, don't worry

File: torch/csrc/Module.cpp (hunk 2 above)
Comment (lqf96) [reply torch/csrc/Module.cpp:289]: @pbelevich I actually submitted a PR to pybind11 a few weeks ago to make pybind11 not assign an empty docstring when all docstring-related options are disabled, however even with that change the `strdup` below is still needed, let alone that change will need time to make its way to pybind11's next release and FB's internal repo...

File: c10/core/GeneratorImpl.h [hunk 3]
```diff
@@ -81,18 +80,9 @@ struct C10_API GeneratorImpl : public c10::intrusive_ptr_target {
...
-  }
-
-  inline PyObject* pyobj() const noexcept {
-    return pyobj_;
-  }
-
   protected:
     Device device_;
     DispatchKeySet key_set_;
-    PyObject* pyobj_ = nullptr;
```
Comment (ezyang) [inline c10/core/GeneratorImpl.h:95]: This looks like a pessimization. If you don't cache the pyobj in the Generator struct, then pybind11 has to manually maintain object identity on its own side, and the only way to do that is with a giant hash table mapping, which is slower than just directly storing it on the object in question. This should show up in the profiles.

File: test/cpp_extensions/rng_extension.cpp [hunk 4]
```diff
@@ -47,7 +47,7 @@ Generator createTestCPUGenerator(uint64_t value) {
   return at::make_generator<TestCPUGenerator>(value);
 }
 
-Generator identity(Generator g) {
+Generator& identity(Generator& g) {
```
Comment (ezyang) [inline test/cpp_extensions/rng_extension.cpp:50]: What's going on here?

File: torch/csrc/Module.cpp (hunk 2 above)
Comment (ezyang) [reply torch/csrc/Module.cpp:289]: We bundle pybind11, so we can take these changes immediately. If it substantially simplifies the implementation in this PR (which I think it does), I'd prefer if we first do a submodule update of pybind11 to the latest master, and then use that to base this PR off of. Don't use ghstack for that, ghstack can't do submodule updates. Someone on FB will have to spearhead the process; @pbelevich, do you think you'll have time to do it? Also cc @wanchaol @seemethere who did updates in the recent history.

**Comment by ezyang** [comment]
cc @dzhulgakov who I remember has advocated that we torchbind Generator so we get bindings automatically. This PR ought to be moving us closer in this direction, but I wonder how torchbind deals with docstrings

File: torch/csrc/Module.cpp (hunk 2 above)
Comment (lqf96) [reply torch/csrc/Module.cpp:289]: @ezyang FYI that PR is pybind/pybind11#2745. With that change we should be able to get rid of these four lines. However, line 291-295 is still necessary because pybind11 manages docstrings with `malloc` and `free`, while `doc_str` (backed by a `std::string`) is probably allocated with operator `new`. If we assign `doc_str` to `ml_doc`, the program may crash when the function is being destructed because it's trying to free the memory of `doc_str` in an incompatible way.

Comment (ezyang) [reply torch/csrc/Module.cpp:289]: cc @malfet

File: c10/core/GeneratorImpl.h (hunk 3 above)
Comment (lqf96) [reply c10/core/GeneratorImpl.h:95]: I agree that pybind11 will need to do lookup in an object identity hash table, but isn't that cost always present with every class you bind with pybind11? If so having `pyobj_` doesn't make sense, because even though you can avoid the lookup, pybind11 will still insert the Python object for the `Generator` into that lookup table. You are only going to end up with some minor speedups at the cost of more memory consumption, and since `Generator` is not that frequently used (here I mean not at the scale of `torch.Tensor`, which we probably never want to bind with pybind11) I prefer not to keep `pyobj_` on `Generator`.

File: test/cpp_extensions/rng_extension.cpp (hunk 4 above)
Comment (lqf96) [reply test/cpp_extensions/rng_extension.cpp:50]: In [`test/test_cpp_extensions_aot.py`](https://github.com/pytorch/pytorch/blob/233192be7334ab1dda217472f67d1de5bcb6685b/test/test_cpp_extensions_aot.py#L161-L163) there is the following code:

```py
copy2 = rng_extension.identity(copy)
self.assertEqual(rng_extension.getInstanceCount(), 1)
self.assertEqual(gen, copy2)
```

If we directly return a `Generator`, a separate Python object will be created. Although both of them refer to the same `GeneratorImpl`, they're not the same Python object and `assertEqual` will fail because we didn't implement comparison for `torch.Generator`. OTOH, accepting and returning a reference to the Generator will tell pybind11 to reuse the same Python object for `g`, thus returning the same object and passing the test.

**Comment by lqf96** [comment]
> cc @dzhulgakov who I remember has advocated that we torchbind Generator so we get bindings automatically. This PR ought to be moving us closer in this direction, but I wonder how torchbind deals with docstrings

I also think it's a good idea to migrate `Generator` to TorchBind in the long term. This will pave the way for using `torch.Generator`s in JITed code.

File: c10/core/GeneratorImpl.h (hunk 3 above)
Comment (ezyang) [reply c10/core/GeneratorImpl.h:95]: You wouldn't have the problem you're fixing in https://github.com/pytorch/pytorch/pull/49771#discussion_r554207279 if you used the `pyobj_` field 👀

But yes, everything bound in pybind11 that uses pybind's automatic identity preservation will be slower this way. You might call this a microoptimization... but the microoptimization is easy and non-invasive; really, the burden of proof should be on *not* doing it. (And that's not considering the correctness problem you have)

Comment (lqf96) [reply c10/core/GeneratorImpl.h:95]: Yeah I agree this will be helpful for performance, but another reason I oppose having `pyobj_` is that it violates the dependency between c10/ATen, PyTorch and `libpython`. Ideally `at::Generator` and `c10::GeneratorImpl` should not depend on `libpython` (instead that should be handled by the PyTorch bindings), however having the `pyobj_` field defeats that purpose.

That said, if you are really concerned about the performance, I have a better idea which does not depends on `pyobj_`. While pybind11 maintains a mapping between Python and C++ objects, it is only looked up when we are casting a pointer or a reference to a Python object. AFAIK, this is only triggered for `Generator.set_state` and `identity` (both return the Generator itself). If we change the generator argument from `Generator` to `py::object` for them, then we should be able to avoid the lookup and remove the overhead.

**Comment by lqf96** [comment]
Any update on this PR? Pybind11 v2.6.2 is out and it contains the docstring PR I made a while ago...

File: c10/core/GeneratorImpl.h (hunk 3 above)
Comment (dzhulgakov) [reply c10/core/GeneratorImpl.h:95]: In terms of build dependency, we have the same in TensorImpl: https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1045 - the reason it works is that you can forward-declare PyObject in libtorch but don't really depend on any python headers.

We can do the same for Generator and return pre-looked-up py::object as you noted in your comment. You probably would need to register custom wrapper with pybind to "remove" the py::object identity from the object when the last reference gets deleted (I'm not sure how exactly to do it, checking pybind sources might be useful)

**Comment by ezyang** [comment]
Besides the arguing about pyobj field (I agree with @dzhulgakov's comment, btw), I wasn't sure what to do with this PR given the developments on #49988. I guess the argument you've made in this PR is that Generator is not performance sensitive and that would apply to pybind11 as well, I suppose... as a compromise, I'm OK with letting this PR through (while torch.device will be dead... for now at least) but I really still want object identity done via `pyobj_` field. Sorry, I make the rules.

**Comment by lqf96** [comment]
@ezyang Let me try the solution I mentioned in https://github.com/pytorch/pytorch/pull/49771#discussion_r554673846, which does not require `pyobj_` and also avoid the lookup. If you are still unsatisfied with that, I can add `pyobj_` back nonetheless. As for performance, I'll add [`noconvert`](https://pybind11.readthedocs.io/en/stable/advanced/functions.html#non-converting-arguments) to `py::args` to reduce the overhead of argument conversion, although performance of pybind11 bindings will remain a long-standing problem.

**Comment by lqf96** [comment]
@ezyang @dzhulgakov This is ready for review again. Let me know if you're OK with the new approach for `set_state` and `manual_seed` that avoids `PyObject` lookup.

**Comment by edward-io** [comment]
Can we revisit this PR? It would be hugely beneficial to checkpoint objects that contain generators, like dataloaders and samplers. cc @ezyang

**Comment by ezyang** [comment]
IMO we should just directly add Pickle support to generator without also refactoring the bindings...
